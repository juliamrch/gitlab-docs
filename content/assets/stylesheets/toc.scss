---
version: 10
---

@import "variables";

$base-nav-link-padding: .75rem;
$sm-nav-link-padding: 1.25rem;
$nav-link-font-size: 14px;

.clear {
  clear: both;
}

// ToC styles
.table-of-contents {
  position: relative;
  padding: 16px 8px;
  margin-top: 60px;

  li {
    list-style: none;
  }

  .nav-link {
    font-size: $nav-link-font-size;
    color: $toc-link-color;
    border-radius: 0;
    border-left: 3px solid $gds-indigo-100;

    &:hover {
      color: $link-color-nav-hover;
      text-decoration: underline;
      background-color: transparent;
      box-shadow: none;
    }

    &.active {
      color: $link-color-nav;
      background-color: transparent;
      border-color: $gds-indigo-800;

      &:hover, &:only-child {
        color: $link-color-nav-hover;
        background-color: transparent;
      }
    }
  }
}

.markdown-toc {
  padding-left: 16px;
}

.nav-pills {
  flex-wrap: nowrap;
}

// not wide enough to show quick nav and toc
@media(max-width:1099px) {
  .doc-nav {
    position: relative;
    min-height: 34px;
  }

  .table-of-contents {
    margin: 0;
    padding: 0;

    .nav-link {
      border-left: 0;
      padding-left: $sm-nav-link-padding;
    }

    h4 {
      font-size: $nav-link-font-size;
    }
  }

  .sm-collapsing {
    height: 0;
    overflow: hidden;
    transition: height .2s ease;
  }

  .sm-collapsed {
    display: none;
  }

  .main.class {
    float: none;
    width: inherit;
    max-width: 900px;
    margin: auto 1% auto 50px;
  }

  .toc-lg {
    display: none;
  }

  .toc-sm {
    display: block;
  }
}

// wide enough to show toc but not quick nav
@media(min-width:1100px) {
  .header {
    top: 0;
    left: 0;
  }

  .doc-nav {
    position: fixed;
    top: 0;
    right: 8px;
    margin-top: 55px;
    overflow: scroll;
    height: 100%;

    &.toc-no-breadcrumbs {
      padding-top: 6px;
    }
  }

  .toc-lg {
    display: block;
  }

  .toc-sm {
    display: none;
  }

  .toc-separator {
    border-top: 1px solid $gds-gray-200;
    padding-top: 1rem;
    margin-top: 1rem;
  }

  // NOTE: This only allows for 10 levels.
  // This should be acceptable since levels are determined by the `<h1-6>` which is less
  // than 10 anyways...
  @for $i from 0 to 10 {
    .nav-link.toc-level-#{$i} {
      padding-left: $base-nav-link-padding * ($i + 1);
    }
  }
}

@media(min-width:1100px) and (max-width:1600px) {
  $doc-nav-width: 19vw;

  .main.class.has-toc {
    width: 64vw;
    max-width: 64vw;
    float: left;
  }

  .doc-nav {
    width: $doc-nav-width;
  }
}

// wide enough to show quick nav and toc
@media(min-width:1601px) {
  $doc-nav-desired-width: 0;
  // Get around the scroll bar
  $doc-nav-width: 21%;

  .main.class.has-toc {
    width: 64vw;
    max-width: 64vw;
    margin: auto 13%;
  }

  .doc-nav {
    width: $doc-nav-width;
  }
}
