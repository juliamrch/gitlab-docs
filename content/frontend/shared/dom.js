/* global $ */

/**
 * Returns outerHeight of element **even if it's hidden**
 *
 * NOTE: Uses jQuery because there is no trivial way to do this in
 * vaniall JS, and it's nice that jQuery has a reliable out-of-the-box
 * solution.
 *
 * @param {Element} el
 */
export const getOuterHeight = (el) => $(el).outerHeight();

/**
 * Find the first child of the given element with the given tag name
 *
 * @param {Element} el
 * @param {String} tagName
 * @returns {Element | null} Returns first child that matches the given tagName (or null if not found)
 */
export const findChildByTagName = (el, tagName) =>
  Array.from(el.childNodes).find((x) => x.tagName === tagName.toUpperCase());
