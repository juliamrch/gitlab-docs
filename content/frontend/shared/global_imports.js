import jquery from 'jquery';
import 'bootstrap';

window.$ = jquery;
