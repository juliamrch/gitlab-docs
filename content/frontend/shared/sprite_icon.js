/**
 * Sprite helper adapted from
 * https://gitlab.com/gitlab-org/gitlab/-/blob/master/app/assets/javascripts/lib/utils/common_utils.js#L472
 */

export const spriteIcon = (icon, className = '') => {
  const spritePath = '/assets/images/icons.svg';
  const classAttribute = className.length > 0 ? `class="${className}"` : '';

  return `<svg ${classAttribute}><use xlink:href="${spritePath}#${icon}" /></svg>`;
};
