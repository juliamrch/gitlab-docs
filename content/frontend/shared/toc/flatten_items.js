// eslint-disable-next-line import/prefer-default-export
export const flattenItems = (items = [], level = 0) => {
  if (!items || !items.length) {
    return items;
  }

  return items.reduce(
    (acc, { items: children, ...item }) =>
      acc.concat([{ ...item, level }]).concat(flattenItems(children, level + 1)),
    [],
  );
};
