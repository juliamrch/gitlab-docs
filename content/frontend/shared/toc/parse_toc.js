/* eslint-disable import/prefer-default-export */
import { findChildByTagName } from '../dom';

const TAG_LI = 'LI';
const TAG_A = 'A';
const TAG_UL = 'UL';

/**
 * Parses the given HTML Table of Contents into a data structure
 *
 * ```
 * type Item = { text: String, href: String, id: String, items: Item[] }
 *
 * parseTOC: Element => Item[]
 * ```
 *
 * @param {Element} menu Parent <ul> element
 */
export const parseTOC = (menu) => {
  const items = [];

  if (!menu) {
    return items;
  }

  menu.childNodes.forEach((li) => {
    if (li.tagName !== TAG_LI) {
      return;
    }

    const link = findChildByTagName(li, TAG_A);
    const subMenu = findChildByTagName(li, TAG_UL);

    if (!link) {
      return;
    }

    const item = {
      text: link.textContent,
      href: link.getAttribute('href'),
      id: link.id,
      items: parseTOC(subMenu),
    };

    items.push(item);
  });

  return items;
};
