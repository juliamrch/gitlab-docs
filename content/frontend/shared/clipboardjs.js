/* global $ */

import { spriteIcon } from './sprite_icon';

const createCopyButton = () => {
  const button = document.createElement('button');
  button.className = 'clip-btn';
  button.title = 'Click to copy';
  button.dataset.selector = 'true';
  button.innerHTML = spriteIcon('copy-to-clipboard', 'gl-icon ml-1 mr-1 s16');
  return button;
};

window.addEventListener('load', () => {
  document.querySelectorAll('pre').forEach((element) => {
    const button = createCopyButton();
    element.appendChild(button);
  });

  $('button').tooltip({
    trigger: 'click',
    placement: 'left',
  });

  function setTooltip(btn, message) {
    $(btn).tooltip('hide').attr('data-original-title', message).tooltip('show');
  }

  function hideTooltip(btn) {
    setTimeout(function hide() {
      $(btn).tooltip('hide');
    }, 1000);
  }

  document.querySelectorAll('.clip-btn').forEach((button) => {
    button.addEventListener('click', (event) => {
      const clickedButton = event.target;

      try {
        navigator.clipboard.writeText(clickedButton.previousElementSibling.innerText);
        setTooltip(clickedButton, 'Copied!');
        hideTooltip(clickedButton);
      } catch {
        setTooltip(clickedButton, 'Failed!');
        hideTooltip(clickedButton);
      }
    });
  });
});
