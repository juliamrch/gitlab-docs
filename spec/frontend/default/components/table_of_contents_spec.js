import { shallowMount, mount } from '@vue/test-utils';
import TableOfContents from '../../../../content/frontend/default/components/table_of_contents.vue';
import * as dom from '../../../../content/frontend/shared/dom';
import { flattenItems } from '../../../../content/frontend/shared/toc/flatten_items';
import { createExampleToc } from '../../shared/toc_helper';

const TEST_ITEMS = createExampleToc();
const TEST_HELP_AND_FEEDBACK_ID = 'test-help-and-feedback';

describe('frontend/default/components/table_of_contents', () => {
  let wrapper;

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  beforeEach(() => {
    // jquery is not available in Jest yet so we need to mock this method
    jest.spyOn(dom, 'getOuterHeight').mockReturnValue(100);
  });

  const createComponent = (props = {}, mountFn = shallowMount) => {
    wrapper = mountFn(TableOfContents, {
      propsData: {
        items: TEST_ITEMS,
        helpAndFeedbackId: TEST_HELP_AND_FEEDBACK_ID,
        ...props,
      },
    });
  };

  const findCollapseButton = () => wrapper.find('[data-testid="collapse"]');
  const findCollapseIcon = () => findCollapseButton().find('i');
  const findCollapsibleContainer = () => wrapper.find('[data-testid="container"]');
  const findMainList = () => wrapper.find('[data-testid="main-list"]');
  const findMainListItems = () => findMainList().props('items');
  const clickCollapseButton = () => findCollapseButton().trigger('click');

  const expectCollapsed = (isCollapsed = true) => {
    expect(findCollapseButton().attributes('aria-expanded')).toBe(isCollapsed ? undefined : 'true');
    expect(findCollapsibleContainer().props('isCollapsed')).toBe(isCollapsed);
    expect(findCollapseIcon().classes(isCollapsed ? 'fa-angle-right' : 'fa-angle-down')).toBe(true);
  };

  it('matches snapshot', () => {
    createComponent({ hasHelpAndFeedback: true }, mount);
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('with hasHelpAndFeedback', () => {
    beforeEach(() => {
      createComponent({ hasHelpAndFeedback: true });
    });

    it('appends help and feedback item', () => {
      expect(findMainListItems()).toEqual(
        flattenItems(
          TEST_ITEMS.concat([
            {
              href: `#${TEST_HELP_AND_FEEDBACK_ID}`,
              id: null,
              items: [],
              text: 'Help and feedback',
              withSeparator: true,
            },
          ]),
        ),
      );
    });
  });

  describe('default', () => {
    beforeEach(() => {
      createComponent({}, mount);
    });

    it('renders toc list', () => {
      expect(findMainListItems()).toEqual(flattenItems(TEST_ITEMS));
    });

    it('is initially uncollapsed', () => {
      expectCollapsed(false);
    });

    describe('when collapse button is pressed', () => {
      beforeEach(() => {
        clickCollapseButton();
      });

      it('starts expanding', () => {
        expect(findCollapsibleContainer().classes('sm-collapsing')).toBe(true);
      });

      it('immediately updates collapse status', () => {
        expectCollapsed(true);
      });

      it('when button pressed again, nothing happens because in the middle of collapsing', () => {
        clickCollapseButton();

        return wrapper.vm.$nextTick(() => {
          expect(findCollapsibleContainer().classes('sm-collapsing')).toBe(true);
          expectCollapsed(true);
        });
      });
    });
  });
});
