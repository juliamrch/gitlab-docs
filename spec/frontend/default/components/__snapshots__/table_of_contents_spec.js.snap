// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frontend/default/components/table_of_contents matches snapshot 1`] = `
<div
  class="table-of-contents-container position-relative"
  id="markdown-toc"
>
  <div
    class="table-of-contents"
  >
    <h4
      class="border-0 toc-sm d-lg-none"
    >
      <a
        aria-controls="markdown-toc"
        aria-expanded="true"
        class="d-flex text-decoration-none"
        data-testid="collapse"
        href="#"
        role="button"
      >
        <i
          class="fa d-flex align-items-center justify-content-center mr-1 gl-w-3 fa-angle-down"
          role="presentation"
        />
        
        On this page
      
      </a>
    </h4>
     
    <h4
      class="border-0 gl-font-base font-weight-bold toc-lg"
    >
      On this page
    </h4>
     
    <div
      class=""
      data-testid="container"
    >
      <ul
        class="nav nav-pills flex-column my-0"
        data-testid="main-list"
      >
        <li>
          <a
            class="nav-link d-block toc-level-0"
            href="#lorem"
            id="Lorem-anchor"
          >
            Lorem
          </a>
           
          <!---->
        </li>
        <li>
          <a
            class="nav-link d-block toc-level-1"
            href="#lorem-"
            id="Lorem 2-anchor"
          >
            Lorem 2
          </a>
           
          <!---->
        </li>
        <li>
          <a
            class="nav-link d-block toc-level-0"
            href="#ipsum"
            id="Ipsum-anchor"
          >
            Ipsum
          </a>
           
          <!---->
        </li>
        <li>
          <a
            class="nav-link d-block toc-level-1"
            href="#dolar"
            id="Dolar-anchor"
          >
            Dolar
          </a>
           
          <!---->
        </li>
        <li>
          <a
            class="nav-link d-block toc-level-2"
            href="#sit"
            id="Sit-anchor"
          >
            Sit
          </a>
           
          <!---->
        </li>
        <li>
          <a
            class="nav-link d-block toc-level-2"
            href="#amit"
            id="Amit-anchor"
          >
            Amit
          </a>
           
          <!---->
        </li>
        <li>
          <a
            class="nav-link d-block toc-level-2"
            href="#test"
            id="Test-anchor"
          >
            Test
          </a>
           
          <!---->
        </li>
        <li>
          <a
            class="nav-link d-block toc-level-0"
            href="#test-help-and-feedback"
          >
            Help and feedback
          </a>
           
          <!---->
        </li>
      </ul>
    </div>
  </div>
</div>
`;
